### PR TITLE
LC-439: support non integer join keys in DbConnector

### PR DESCRIPTION
**Pull request by Jozurf**

### CODE REVIEW

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java [hunk 1]
```diff
@@ -253,18 +252,20 @@ private void iterateOtherSQL(ResultSet rs2, String[] columns2, Document doc, Int
...
     }
-
     do {
       // Convert to do-while I think we can avoid the rs2.previous() call.
-      // TODO: support non INT primary key
-      int otherJoinId = rs2.getInt(joinField);
-      if (otherJoinId < joinId) {
+      Object otherJoinId = rs2.getObject(joinField);
+
+      if (!(otherJoinId instanceof Comparable) || !(joinId instanceof Comparable) || !otherJoinId.getClass().equals(joinId.getClass())) {
```
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:259]: is there a way to do this check only once -- somewhere outside of the do/while?

Comment (Jozurf) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java]: yes. Resolved in [cfbc0e0](https://github.com/kmwtechnology/lucille/pull/184/commits/cfbc0e07bb2e737c6c86c682d3dbb9e8d9241c87)

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java [hunk 2]
```diff
@@ -255,16 +254,14 @@ private void iterateOtherSQL(ResultSet rs2, String[] columns2, Document doc, Int
     }
 
     do {
-      // Convert to do-while I think we can avoid the rs2.previous() call.
-      // TODO: support non INT primary key
-      int otherJoinId = rs2.getInt(joinField);
-      if (otherJoinId < joinId) {
+      Object otherJoinId = rs2.getObject(joinField);
+      // will throw error if different class, if either of objects are null or not Comparable
+      int comparison = ((Comparable) otherJoinId).compareTo(joinId);
```
Comment (spencersolomon6) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:259]: Would probably be helpful to catch each of these specific exceptions and then throw a ConnectorException with the original exception and a message explaining what went wrong.

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java [hunk 3]
```diff
@@ -345,6 +346,115 @@ public void testJoiningDatabaseConnector() throws Exception {
...
+    ArrayList<String> otherJoinFields = new ArrayList<>();
+    otherJoinFields.add("metadata");
+    configValues.put("otherJoinFields", otherJoinFields);
+    // create a config object off that map
+    Config config = ConfigFactory.parseMap(configValues);
+
+    // create the connector with the config
+    DatabaseConnector connector = new DatabaseConnector(config);
+    // run the connector, expect to throw non-comparable error
+    assertThrows("Other join field or joinId is not comparable.",ConnectorException.class, () -> connector.execute(publisher));
```
Comment (spencersolomon6) [inline lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:414]: Where is this exception/text in the code? I can't seem to find it.

**Review by spencersolomon6 — CHANGES_REQUESTED** [review]

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java (hunk 2 above)
Comment (Jozurf) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java]: implemented catching for all exceptions in [91eec79](https://github.com/kmwtechnology/lucille/pull/184/commits/91eec79718ff3de576700b2d2911d5c1515ea985)

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java (hunk 3 above)
Comment (Jozurf) [reply lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java]: Somehow, the test still passes when error thrown does not have an accompanying message or a different message. Removed the exception text from the test in [91eec79](https://github.com/kmwtechnology/lucille/pull/184/commits/91eec79718ff3de576700b2d2911d5c1515ea985) as we are longer throwing an error with that message.

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java [hunk 4]
```diff
@@ -255,8 +255,14 @@ private void iterateOtherSQL(ResultSet rs2, String[] columns2, Document doc, Obj
...
       Object otherJoinId = rs2.getObject(joinField);
-      // will throw error if different class, if either of objects are null or not Comparable
-      int comparison = ((Comparable) otherJoinId).compareTo(joinId);
+      int comparison;
+      try {
+        comparison = ((Comparable) otherJoinId).compareTo(joinId);
+      } catch (NullPointerException e) {
+        throw new ConnectorException("Either otherJoinId or joinId is null.");
+      } catch (ClassCastException e) {
+        throw new ConnectorException("Either otherJoinId or JoinId class type prevents them from being compared.");
```
Comment (spencersolomon6) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnector.java:264]: Include the original Stacktrace in these

`new ConnectorException("...", e);`

**Review by spencersolomon6 — APPROVED** [review]

File: lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java [hunk 5]
```diff
@@ -44,6 +48,7 @@ public class DatabaseConnectorTest {
   @Rule
   public final DBTestHelper dbHelper = new DBTestHelper("org.h2.Driver", "jdbc:h2:mem:test", "",
       "", "db-test-start.sql", "db-test-end.sql");
+  private static TimeZone originalTimeZone;
```
Comment (rseitz) [inline lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java:51]: i think you can make this final and initialize it to TimeZone.getDefault() all on this line. Then you don't need to do it in setUpTimeZone()

Comment (Jozurf) [reply lucille-core/src/test/java/com/kmwllc/lucille/connector/jdbc/DatabaseConnectorTest.java]: done in [26fc6fc](https://github.com/kmwtechnology/lucille/pull/184/commits/26fc6fc18d34beada13380cc47717f51ed6b7d9d)

**Review by rseitz — APPROVED** [review]

**Pull request merged**